### PR TITLE
Add support for bokeh event callbacks

### PR DESCRIPTION
**Pull request by philippjfr**
Adds support for the upcoming event callbacks in bokeh, currently being developed [here](https://github.com/jlstevens/bokeh/pull/4).

Summary of how it works
===================

1. Check if any of the ``skip_conditions`` apply, if so, skip

2. Grab all the requested ``attributes`` from the models and accumulate them in data

3. Check if there is a ``Comm`` instantiated otherwise open it and store it in a global registry

4. Check if there is a ``comm_state`` for the ``Comm`` otherwise make one defining a ``event_buffer``, a ``blocked`` flag and a ``time``

5. Check whether the comm_state is blocked or has timed out (i.e. ``Date.now()>comm_state.time+timeout``):

   a) if it's blocked put data along with ``cb_obj.event.event_name`` on top of the ``event_buffer``

   b) if it's not blocked also prepend to ``event_buffer`` but also call ``process_event`` with small timeout. The small timeout acts as a debouncing mechanism as new events can be pushed ahead of the original event in the event queue. After process_events is called set the ``comm_state`` to blocked

6. ``process_events`` will iterate over the ``event_buffer`` and filter out older duplicated events. This ensures only the latest event of each type is processed (e.g. to ensure ``panstart``, ``pan`` and ``panend`` are all processed). (Can probably be simplified)

7. In python everything does its thing and sends back a message either acknowledging execution or failure. This message also contains the ``comm_id``, with which the ``comm_state`` is looked up. If there is something on the ``event_buffer`` process it and keep the ``Comm`` blocked, otherwise unblock.

ToDo:

- [x] Update docstring
- [x] Finalize naming for Callback attributes
- [ ] Add tests to ensure callbacks reference correct handles/models
- [ ] Add documentation about defining custom Stream/Callbacks

### CODE REVIEW

File: holoviews/plotting/bokeh/callbacks.py [hunk 1]
```diff
@@ -92,19 +104,41 @@ class Callback(object):
     attributes = {}
 
     js_callback = """
+        function unique_events(events) {{
```
Comment (jlstevens) [inline holoviews/plotting/bokeh/callbacks.py]: As far as I can tell this is a self-contained function that could be a utility. Would be good to have a docstring i.e a comment line saying that it does and the same for ``process_events`` below. At some point we will probably want the concept of Javascript 'utilities' that you can just include where needed.

That said, I just noticed ``comm_state`` is referenced in this function, can it not be made into an explicit function argument?

Comment (philippjfr) [inline holoviews/plotting/bokeh/callbacks.py]: This is not ready to review. I've got outstanding changes locally.

File: holoviews/plotting/bokeh/callbacks.py [hunk 2]
```diff
@@ -92,19 +104,41 @@ class Callback(object):
...
+                if (!unique_events.includes(event)) {{
+                    unique.unshift(data);
+                    unique_events.push(event);
+                }}
+            }}
+            return unique;
+        }}
+
+        function process_events(comm_state) {{
+            var events = unique_events(comm_state.events);
```
Comment (jlstevens) [inline holoviews/plotting/bokeh/callbacks.py]: So this is at least one place where ``unique_events`` is used and ``comm_state`` is available. You could definitely pass ``comm_state`` to ``unique_events`` here though ``unique_events(comm_state.events, comm_state)`` is a bit redundant (just pass ``comm_state``). The fact that you can access ``comm_state`` at all in ``unique_event`` without explicitly declaring it makes me uncomfortable (but then again, this is Javascript!)

Comment (philippjfr) [inline holoviews/plotting/bokeh/callbacks.py]: As I said above, I would not review this yet.

File: holoviews/plotting/bokeh/element.py [hunk 3]
```diff
@@ -104,6 +103,12 @@ class ElementPlot(BokehPlot, GenericElementPlot):
           * timeout   - Timeout (in ms) for checking whether interactive
                         tool events are still occurring.""")
 
+    sizing_mode = param.ObjectSelector(default='fixed',
+                                       objects=[
+            'fixed', 'scale_width', 'scale_height',
+            'scale_both', 'stretch_both'
+        ], doc="""Defines how the plot scales when resized.""")
+
```
Comment (jlstevens) [inline holoviews/plotting/bokeh/element.py]: Is this relevant to hooking up to events?

Comment (philippjfr) [inline holoviews/plotting/bokeh/element.py]: Please stop reviewing this PR.

File: holoviews/plotting/bokeh/callbacks.py (hunk 2 above)
Comment (jlstevens) [inline holoviews/plotting/bokeh/callbacks.py]: Fair enough - I was just trying to understand the code out of my own curiosity and thought I might as well make some comments while I'm at it.

Comment (philippjfr) [inline holoviews/plotting/bokeh/callbacks.py]: Looking is fine, but there's no point reviewing things I'm still actively working on.

File: holoviews/plotting/bokeh/element.py (hunk 3 above)
Comment (jlstevens) [inline holoviews/plotting/bokeh/element.py]: Why not just work on a branch if it is totally WIP? Once you make a PR you make the proposed changes public in a way that invites comment even if the status is WIP.

I would recommend working on a branch and only turning it into a PR when you require some sort of feedback.

Comment (philippjfr) [inline holoviews/plotting/bokeh/element.py]: >I would recommend working on a branch and only turning it into a PR when you require some sort of feedback.

I disagree, I like being able to easily see an overview of my changes on Github and that's exactly what the WIP and ready tags are for.

File: holoviews/plotting/bokeh/callbacks.py (hunk 1 above)
Comment (philippjfr) [inline holoviews/plotting/bokeh/callbacks.py]: Independent of the details of this function a set of JS utilities definitely makes sense. The issue then is that we will really want to have a proper JS package, because polluting the main namespace isn't ideal. All things to consider once we get started writing the widget manager.

File: holoviews/plotting/bokeh/element.py (hunk 3 above)
Comment (jbednar) [inline holoviews/plotting/bokeh/element.py]: But if it's really not ready for someone to look at, in addition to tagging it WIP you should put a sentence in the description saying what you're up to so that other people know what the status us.  Lots of things marked WIP are ready to look at, just not ready to merge.

Comment (jlstevens) [inline holoviews/plotting/bokeh/element.py]: A WIP tag says that the *pull request* is in work in progress which means there is a public conversation going on about a feature that is in progress. That isn't the same as 'please do not comment' and it sounds like that is how you want to define it.

Github defines a [PR](https://help.github.com/articles/about-pull-requests/) as:

> Pull requests let you tell others about changes you've pushed to a repository on GitHub.

It is clearly there for the purposes of communication, even if you are communicating to yourself (e.g to see an overview of your changes). I'm wondering if there is a tag that better conveys what you intended - not only 'work in progress' but 'don't supply feedback yet'.

Comment (philippjfr) [inline holoviews/plotting/bokeh/element.py]: Agreed, I'll add comments, but the usage of WIP across all the projects I've encountered has meant not ready for review, and ``ready`` means ready for review not necessarily merge.

File: holoviews/plotting/bokeh/callbacks.py (hunk 1 above)
Comment (jlstevens) [inline holoviews/plotting/bokeh/callbacks.py]: Agreed.

File: holoviews/plotting/bokeh/element.py (hunk 3 above)
Comment (philippjfr) [inline holoviews/plotting/bokeh/element.py]: A PR is a place for discussion, which is often helpful long before the code is ready to review.

Comment (jlstevens) [inline holoviews/plotting/bokeh/element.py]: Ah, I see Jim agrees with me here. :-)

'Not ready for review' is a bit long, but could just about be a tag? I also agree that there are different conventions across projects which is why a WIP tag by itself is ambiguous.

Comment (jbednar) [inline holoviews/plotting/bokeh/element.py]: What it usually means in the projects I'm familiar with is *definitely* not ready to merge, *probably* not ready for heavyhanded, very detailed comments, and *probably* ready for someone to take a quick look and start giving some feedback.  If that's not the case for a given WIP PR, just declare it in text so that other humans know what to do with it.  No big deal...

Comment (jlstevens) [inline holoviews/plotting/bokeh/element.py]: How about a ``pre-review`` tag if you don't want comment? I'm not sure it replaces the ``WIP`` tag but from the sound of it, it might do.

Comment (philippjfr) [inline holoviews/plotting/bokeh/element.py]: I think high-level feedback on a WIP PR makes sense but not necessarily which arguments should be passed to which functions. I think we should just clarify what the tags mean.

Comment (jlstevens) [inline holoviews/plotting/bokeh/element.py]: The problem with a PR like this is that the description at the top is very brief and nothing worth commenting on. That means to know what the PR is really about I have to look at the code, at which point it becomes very tempting to make comments, even without delving into the exact details i.e just trying to understand what is going on.

That is probably why what Jim suggests makes the most sense - just say you aren't ready to receive comments just yet at the top.

Comment (philippjfr) [inline holoviews/plotting/bokeh/element.py]: >That is probably why what Jim suggests makes the most sense - just say you aren't ready to receive comments just yet at the top.

Agreed, this PR has been around a while and I forgot to update the description. That said I'm still pretty firm on the meanings of ``WIP`` and ``ready``, but without putting that information somewhere I guess no one will know what those conventions are.

Comment (jlstevens) [inline holoviews/plotting/bokeh/element.py]: How about ``WIP`` and ``WIP-discussion``? Then we can have the semantics you want.

Comment (jbednar) [inline holoviews/plotting/bokeh/element.py]: Or `WIP` and `WIP-pre-review`...

Comment (jlstevens) [inline holoviews/plotting/bokeh/element.py]: Then ``WIP`` allows for discussion. :-)

That said, I *believe* you can rename GitHub tags so all existing PRs and issues get renamed at once. Personally, I like Jim's suggestion more but I am happy to be overruled.

Comment (jbednar) [inline holoviews/plotting/bokeh/element.py]: Also, I personally prefer that a WIP PR includes an explicit checklist in the description, stating what the author intends to do before it will be ready, and marking those off as they are done.  That way it's fully ready for review about *some* things, and other things are clearly not ready for review.  This PR doesn't have such a checklist, which is one reason it's hard for other people to know what to do with it.

Comment (jlstevens) [inline holoviews/plotting/bokeh/element.py]: By the way, GitHub supports PR templates which is where this information could go. I would keep it short as a long list of requirements/restrictions/demands would put people off contributing.

Asking for a checklist of items would be a very reasonable thing to do though.

Comment (philippjfr) [inline holoviews/plotting/bokeh/element.py]: I agree, I definitely think we should add a PR and Issue template.

**Comment by philippjfr** [comment]
1. Check if any of the ``skip_conditions`` apply, if so, skip
2. Grab all the requested ``attributes`` from the models and accumulate them in data
3. Check if there is a ``Comm`` instantiated otherwise open it and store it in a global registry
4. Check if there is a ``comm_state`` for the ``Comm`` otherwise make one defining a ``event_buffer``, a ``blocked`` flag and a ``time``
5. Check whether the comm_state is blocked or has timed out (i.e. ``Date.now()>comm_state.time+timeout``):
a) if it's blocked put data along with ``cb_obj.event.event_name`` on top of the ``event_buffer``
b) if it's not blocked also prepend to ``event_buffer`` but also call ``process_event`` with small timeout. The small timeout acts as a debouncing mechanism as new events can be pushed ahead of the original event in the event queue. After process_events is called set the ``comm_state`` to blocked
``process_events`` will iterate over the ``event_buffer`` and filter out older duplicated events. This ensures only the latest event of each type is processed (e.g. to ensure ``panstart``, ``pan`` and ``panend`` are all processed). (Can probably be simplified)

6. In python everything does its thing and sends back a message either acknowledging execution or failure. This message also contains the ``comm_id``, with which the ``comm_state`` is looked up. If there is something on the ``event_buffer`` process it and keep the ``Comm`` blocked, otherwise unblock.

**Comment by philippjfr** [comment]
Ready for review.

File: holoviews/plotting/bokeh/element.py [hunk 4]
```diff
@@ -370,9 +374,11 @@ def _init_plot(self, key, element, plots, ranges=None):
             properties['toolbar_location'] = self.toolbar
 
         properties['webgl'] = Store.renderers[self.renderer.backend].webgl
-        return bokeh.plotting.Figure(x_axis_type=x_axis_type,
-                                     y_axis_type=y_axis_type, title=title,
-                                     **properties)
+        with warnings.catch_warnings():
```
Comment (jlstevens) [inline holoviews/plotting/bokeh/element.py]: Probably worth mentioning what warning you are catching and why it doesn't matter that you are suppressing it.

File: holoviews/plotting/bokeh/callbacks.py [hunk 5]
```diff
@@ -53,35 +55,55 @@ def attributes_js(attributes, handles):
...
+    * models     :  Defines which bokeh models the callback will be
+                    attached on referencing the model by its key in
+                    the plots handles, e.g. this could be the x_range,
+                    y_range, plot, a plotting tool or any other
+                    bokeh mode.
 
-    * attributes :  The attributes define which attributes to send
+    * extra_models: Any additional models available in handles which
+                    should be made available in the namespace of the
+                    objects.
```
Comment (jlstevens) [inline holoviews/plotting/bokeh/callbacks.py]: Giving an example of why you might want this would help here. E.g 'For instance, this can be used to check which tools are currently active'.

File: holoviews/plotting/bokeh/callbacks.py [hunk 6]
```diff
@@ -53,35 +55,55 @@ def attributes_js(attributes, handles):
...
                     you can supply {'x': 'x_range.attributes.start'}.
                     Additionally certain handles additionally make the
                     cb_data and cb_obj variables available containing
                     additional information about the event.
 
+    * skip        : Conditions when the Callback should be skipped
+                    specified as a list of valid JS expressions, which
+                    can reference models requested by the callback,
+                    e.g. ['pan.attributes.active'] would skip the
+                    callback if the pan tool is active.
```
Comment (jlstevens) [inline holoviews/plotting/bokeh/callbacks.py]: If the models references in the skip conditions need to be in ``extra_models`` it would be helpful to say so.

File: holoviews/plotting/bokeh/callbacks.py (hunk 1 above)
Comment (jlstevens) [inline holoviews/plotting/bokeh/callbacks.py]: Ok, the signature now looks fine and making JS utilities is for some other day.

File: holoviews/plotting/bokeh/callbacks.py [hunk 7]
```diff
@@ -129,35 +178,42 @@ class Callback(object):
...
-               var comm_id = comm_state.event["comm_id"]
-               var comm = HoloViewsWidget.comms[comm_id];
-               comm.send(comm_state.event);
-            }}
-            comm_state.event = undefined;
-        }}
-
-        timeout = comm_state.timeout + {timeout};
+        // Add current event to queue and process queue if not blocked
+        event_name = cb_obj.event ? cb_obj.event.event_name : undefined
```
Comment (jlstevens) [inline holoviews/plotting/bokeh/callbacks.py]: I think this line is expected to change - I've been requested to change how this works in Bokeh.

File: holoviews/plotting/bokeh/callbacks.py [hunk 8]
```diff
@@ -215,13 +279,18 @@ def _filter_msg(self, msg, ids):
...
             filtered_msg = self._filter_msg(msg, ids)
             processed_msg = self._process_msg(filtered_msg)
             if not processed_msg:
                 continue
             stream.update(trigger=False, **processed_msg)
+            stream._metadata = {h: {'id': hid, 'events': self.events}
+                                for h, hid in metadata.items()}
         Stream.trigger(self.streams)
+        for stream in self.streams:
+            stream._metadata = None
```
Comment (jlstevens) [inline holoviews/plotting/bokeh/callbacks.py]: I would consider just declaring ``_metadata`` to be an empty dict by default. I'm assuming this loop is to reset...

File: holoviews/streams.py [hunk 9]
```diff
@@ -124,6 +129,8 @@ def __init__(self, preprocessors=[], source=None, subscribers=[], **params):
         self.subscribers = subscribers
         self.preprocessors = preprocessors
         self._hidden_subscribers = []
+        self._metadata = None
```
Comment (jlstevens) [inline holoviews/streams.py]: Should probably be an empty dict and I would add a comment what it is for - right now it is about stopping loops.

File: holoviews/plotting/bokeh/callbacks.py (hunk 6 above)
Comment (philippjfr) [inline holoviews/plotting/bokeh/callbacks.py]: That's what this line was about:

>which can reference models requested by the callback,

but I'll try to clarify.

File: holoviews/plotting/bokeh/callbacks.py (hunk 7 above)
Comment (philippjfr) [inline holoviews/plotting/bokeh/callbacks.py]: Okay, can fix that in a follow-up PR when you've made that change.

File: holoviews/streams.py (hunk 9 above)
Comment (philippjfr) [inline holoviews/streams.py]: Agreed, definitely needs a comment.

File: holoviews/plotting/bokeh/callbacks.py [hunk 10]
```diff
@@ -272,54 +340,109 @@ def set_customjs(self, handle, references):
...
-        if id(handle.callback) in self._callbacks:
-            cb = self._callbacks[id(handle.callback)]
+        conditions = ["%s" % cond for cond in self.skip]
+        conditional = ''
+        if conditions:
+            conditional = 'if (%s) { return };\n' % (' || '.join(conditions))
+        data = "var data = {};\n"
+        code = conditional + data + attributes + self.code + self_callback
+
+        js_callback = CustomJS(args=references, code=code)
```
Comment (jlstevens) [inline holoviews/plotting/bokeh/callbacks.py]: I would consider factoring out a method that returns a ``CustomJS`` so all the Javascript is kept separate (probably including the call to ``attributes_js``). The method could be called something like ``get_customjs``...

File: holoviews/plotting/bokeh/callbacks.py [hunk 11]
```diff
@@ -272,54 +340,109 @@ def set_customjs(self, handle, references):
...
+        cb = None
+        if id(handle) in self._callbacks:
```
Comment (jlstevens) [inline holoviews/plotting/bokeh/callbacks.py]: Use of ``id`` here probably warrants a comment about what it is for...

File: holoviews/streams.py [hunk 12]
```diff
@@ -72,6 +72,10 @@ class Stream(param.Parameterized):
 
     Streams may have one or more subscribers which are callables passed
     the parameter dictionary when the trigger classmethod is called.
+
+    Depending on the plotting backend certain streams may interactively
+    subscribe to events and changes by the plotting backend. To disable
+    this behavior instantiate the Stream with interactive=False.
```
Comment (jlstevens) [inline holoviews/streams.py]: I wish there were a better name for this concept by ``interactive`` is the best I can think of right now...

Comment (jlstevens) [inline holoviews/streams.py]: How about ``live``?

Comment (jlstevens) [inline holoviews/streams.py]: Then we can pretend we are doing 'live streams' ;-p

Comment (philippjfr) [inline holoviews/streams.py]: Prefer ``interactive`` tbh.

Comment (jlstevens) [inline holoviews/streams.py]: @jbednar Any thoughts?

The concept here is you might want a ``PositionXY`` stream you want to manage yourself and you don't want the bokeh backend to try to supply you values. By default Bokeh supplies values but this option allows you to switch it off if you want (e.g you want to use streams in the same way between bokeh an matplotlib).

File: holoviews/plotting/bokeh/callbacks.py [hunk 13]
```diff
@@ -215,13 +279,18 @@ def _filter_msg(self, msg, ids):
 
     def on_msg(self, msg):
         for stream in self.streams:
-            ids = self.handle_ids[stream]
+            metadata = self.handle_ids[stream]
```
Comment (jlstevens) [inline holoviews/plotting/bokeh/callbacks.py]: I think I would just call this ``handle_ids`` here.

File: holoviews/plotting/bokeh/callbacks.py [hunk 14]
```diff
@@ -215,13 +279,18 @@ def _filter_msg(self, msg, ids):
...
+            ids = list(metadata.values())
             filtered_msg = self._filter_msg(msg, ids)
             processed_msg = self._process_msg(filtered_msg)
             if not processed_msg:
                 continue
             stream.update(trigger=False, **processed_msg)
+            stream._metadata = {h: {'id': hid, 'events': self.events}
+                                for h, hid in metadata.items()}
```
Comment (jlstevens) [inline holoviews/plotting/bokeh/callbacks.py]: ... and use ``... for h, hid in handle_ids.items()`` here...

Comment (jlstevens) [inline holoviews/plotting/bokeh/callbacks.py]: Also, if you expect to use ``_metadata`` for anything else, then you'll need a key e.g ``'prevent_loop_ids'`` (or similar).

File: holoviews/streams.py (hunk 12 above)
Comment (jbednar) [inline holoviews/streams.py]: automatic? autoupdate? tied?

Comment (philippjfr) [inline holoviews/streams.py]: ``linked``?

Comment (jbednar) [inline holoviews/streams.py]: Sure, linked.

Comment (jlstevens) [inline holoviews/streams.py]: ``linked`` isn't bad

**Comment by jlstevens** [comment]
There are a few minor outstanding things - just tell me when you think it is ready to merge and I'll merge (as long as the tests are passing).

**Comment by philippjfr** [comment]
I've addressed all your comments except for the change you are still going to make in bokeh. It's just that you commented on the line above in some cases.

**Comment by jlstevens** [comment]
Ok, merging.
